pkg/geoipupdate/database/local_file_writer.go

```diff
@@ -0,0 +1,166 @@
+package database
+
+import (
+	"crypto/md5"
+	"fmt"
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+const zeroMD5 = "00000000000000000000000000000000"
+
+//LocalFileDatabaseWriter is a database.Writer that stores the database to the local file system
+type LocalFileDatabaseWriter struct {
+	filePath string
+	lockFile string
+	verbose  bool
+	lock     *flock.Flock
+	oldHash  string
+	swapFile *os.File
+}
+
+//NewLocalFileDatabaseWriter create a new LocalFileDatabaseWriter, creating necessary lock and swap files to protect
+// the database from concurrent writes
+func NewLocalFileDatabaseWriter(filePath string, lockFile string, verbose bool) (*LocalFileDatabaseWriter, error) {
+	dbWriter := &LocalFileDatabaseWriter{
+		filePath: filePath,
+		lockFile: lockFile,
+		verbose:  verbose,
+	}
+	file, err := os.Open(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			dbWriter.oldHash = zeroMD5
+		} else {
+			return nil, errors.Wrap(err, "Received an unexpected error attempting to open file "+filePath)
+		}
+	} else {
+		defer func() {
+			err := file.Close()
+			if err != nil {
+				log.Println(errors.Wrap(err, "Error closing current database file "+filePath))
+			}
+		}()
+		hash := md5.New()
+		if _, err := io.Copy(hash, file); err != nil {
+			return nil, errors.Wrap(err, "Encountered an error while createing hash for file "+filePath)
+		}
+		dbWriter.oldHash = fmt.Sprintf("%x", hash.Sum(nil))
+		if verbose {
+			log.Printf("Calculated MD5 sum for %s: %s", filePath, dbWriter.oldHash)
+		}
+	}
+	if err := dbWriter.lockDirectory(); err != nil {
+		return nil, err
+	}
+
+	dbWriter.swapFile, err = os.Create(fmt.Sprintf("%s.swap", dbWriter.filePath))
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating swap file")
+	}
+
+	return dbWriter, nil
+}
+
+func (writer *LocalFileDatabaseWriter) lockDirectory() error {
```

Comment:
This is using the global lock file rather than the directory. Should we update the name?

Reply:
I'm uncertain that this is a global lock.  I might not fully understand the flock package, but it seems that it is a lock on a single file; if that is the case, writes can happen on any file except for the lock file.  I can still make the name update, but I think it is a bit misleading to call it a global lock.